api/base/serializers.py

```diff
@@ -404,9 +404,17 @@ def to_internal_value(self, data):
             type_ = get_meta_type(self.root.child, request)
         else:
             type_ = get_meta_type(self.root, request)
-
-        if type_ != data:
-            raise api_exceptions.Conflict(detail=('This resource has a type of "{}", but you set the json body\'s type field to "{}". You probably need to change the type field to match the resource\'s type.'.format(type_, data)))
+        if StrictVersion(request.version) != StrictVersion('2.15'):
+            if type_ != data:
+                raise api_exceptions.Conflict(detail=('This resource has a type of "{}", but you set the json body\'s type field to "{}". You probably need to change the type field to match the resource\'s type.'.format(type_, data)))
+        else:
+            kebab_case = str(type_).replace('-', '_')
+            if type_ != data:
+                if kebab_case == data:
+                    type_ = kebab_case
+                    self.context['request'].META.setdefault('warning', 'As of API Version 2.15, all types are now Kebab-case. 2.15 will accept snake_case, but this will be deprecated in future versions.')
```

Comment:
I think we should rethink this logic - post-2.15 you're going to get a 409 with snake_case versions.  I don't think this is adequate warning to give users that we're deprecating versions, as we have several PR's in the queue that are updating versions.  When this PR goes out, it's quite possible that the next version comes shortly after this, and all of a sudden, snake_case does not work with newer versions.  I think we need to have a longer roll-out period. 

For now, I would return only snake-case after 2.15, but before 2.15 still accept snake_case going forward for now.
So I'd change logic to be, if request type case is snake-case version, remind them that it will be deprecated soon (but don't throw error), but if type is just plain wrong throw a 409.  (And I'd want James to be okay with this!)